libtcmu_store.c

```diff
@@ -19,101 +19,628 @@
+
+free_writecmd:
+	write_verify_free_writecmd(writecmd);
+free_readcmd:
+	write_verify_free_readcmd(readcmd);
+out:
+	return ret;
+}
+
+/* async compare_and_write */
+
+struct tcmu_caw_state {
+	off_t off;
+	ssize_t requested;
+	struct tcmulib_cmd *origcmd;
+};
+
+static struct tcmulib_cmd *
+caw_init_readcmd(struct tcmulib_cmd *origcmd,
+		 off_t off, ssize_t length, size_t iov_cnt)
+{
+	struct iovec *iov;
+	struct tcmulib_cmd *readcmd;
+	struct tcmu_caw_state *state;
+
+	state = calloc(1, sizeof(*state));
+	if (!state)
+		goto out;
+	readcmd = calloc(1, sizeof(*readcmd));
+	if (!readcmd)
+		goto free_state;
+	iov = alloc_and_assign_iovec(readcmd, length, iov_cnt);
+	if (!iov)
+		goto free_cmd;
+
+	/* multi-op state maintainance */
+	state->off = off;
+	state->requested = length;
+	state->origcmd = origcmd;
+
+	readcmd->cmdstate = state;
+	return readcmd;
+
+free_cmd:
+	free(readcmd);
+free_state:
+	free(state);
+out:
+	return NULL;
+}
+
+static void caw_free_readcmd(struct tcmulib_cmd *readcmd)
+{
+	struct tcmu_caw_state *state = readcmd->cmdstate;
+
+	free_iovec(readcmd);
+	free(readcmd);
+	free(state);
+}
+
+static void call_store_write_cbk(struct tcmu_device *dev,
+				 struct tcmulib_cmd *tcmulib_cmd, int ret);
+
+static void call_store_caw_read_cbk(struct tcmu_device *dev,
+				    struct tcmulib_cmd *readcmd, int ret)
+{
+	uint32_t cmp_offset;
+	struct tcmu_call_stub stub;
+	struct tcmu_caw_state *state = readcmd->cmdstate;
+	struct tcmulib_cmd *origcmd = state->origcmd;
+	uint8_t *sense = origcmd->sense_buf;
+	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
+	struct tcmur_handler *store = handler->hm_private;
+
+	/* read failed - bail out */
+	if (ret != SAM_STAT_GOOD)
+		goto finish_err;
+
+	/* verify failed - bail out */
+	cmp_offset = tcmu_compare_with_iovec(readcmd->iovec->iov_base,
+					     origcmd->iovec, state->requested);
+	if (cmp_offset == -1) {
+		ret = tcmu_set_sense_data(sense, MISCOMPARE,
+					  ASC_MISCOMPARE_DURING_VERIFY_OPERATION,
+					  &cmp_offset);
+		goto finish_err;
+	}
+
+	/* perform write */
+	tcmu_seek_in_iovec(origcmd->iovec, state->requested);
+	stub.sop = TCMU_STORE_OP_WRITE;
+	stub.callout_cbk = call_store_write_cbk;
+
+	stub.u.rw.exec = store->write;
+	stub.u.rw.iov = origcmd->iovec;
+	stub.u.rw.iov_cnt = origcmd->iov_cnt;
+	stub.u.rw.off = state->off;
+
+	ret = async_call_command(dev, origcmd, &stub);
+	if (ret != TCMU_ASYNC_HANDLED)
+		goto finish_err;
+
+	caw_free_readcmd(readcmd);
+	return;
+
+finish_err:
+	caw_free_readcmd(readcmd);
+	tcmu_command_finish(dev, origcmd, ret, true);
+}
+
+static int call_store_caw(struct tcmu_device *dev,
+			  struct tcmur_handler *store,
+			  struct tcmulib_cmd *tcmulib_cmd,
+			  struct iovec *iovec, size_t iov_cnt, off_t off)
+{
+	int ret;
+	struct tcmu_call_stub stub;
+	struct tcmulib_cmd *readcmd;
+	uint8_t *sense = tcmulib_cmd->sense_buf;
+	ssize_t half = (tcmu_iovec_length(iovec, iov_cnt)) / 2;
+
+	ret = errno_to_sam_status(-ENOMEM, sense);
+
```

Comment:
Ummm.. looks like `file_optimal` too does not lock around the compound operations. Also, this would mean introducing locks in `read`, `write` and `flush` paths.